torchci/scripts/check_alerts.py

```diff
@@ -0,0 +1,296 @@
+REPO_NAME = "test-infra"
+failure_label = "pytorch-alert"
+labels = [failure_label]
+headers = {"Authorization": f"token {os.environ.get('GITHUB_TOKEN')}"}
+
+
+CREATE_ISSUE_URL = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues"
+UPDATE_ISSUE_URL = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues/"
+
+GRAPHQL_URL = "https://api.github.com/graphql"
+
+
+class JobStatus:
+    job_name: str = ""
+    jobs: List[Any] = []
+    current_status: str = None
+    job_statuses: List[Any] = []
+    failure_chain: List[Any] = []
+    flaky_jobs: List[Any] = []
+
+    def __init__(self, job_name: str, job_statuses: List[Any]):
+        self.job_name = job_name
+        self.job_statuses = job_statuses
+
+        filtered_statuses = list(filter(is_job_not_pending_or_skipped, job_statuses))
+        self.current_status = self.get_current_status(filtered_statuses)
+        self.failure_chain = self.get_most_recent_failure_chain(filtered_statuses)
+        self.flaky_jobs = self.get_flaky_jobs(filtered_statuses)
+
+    def get_current_status(self, job_statuses: List[Any]) -> str:
+        return job_statuses[0] if len(job_statuses) > 0 else None
+
+    # Returns a dict of failureCaptures -> List[Jobs]
+    def get_unique_failures(self, job_statuses: List[Any]) -> Dict[str, List[Any]]:
+        failures = {}
+        for job in job_statuses:
+            if job["conclusion"] == "failure":
+                found_similar_failure = False
+                if "failureCaptures" not in job:
+                    if "unclassified" not in failures:
+                        failures["unclassified"] = []
+                    failures["unclassified"].append(job)
+                    continue
+
+                for failure in failures:
+                    seq = SequenceMatcher(None, job["failureCaptures"], failure)
+                    if seq.ratio() > 0.75:
+                        failures[failure].append(job)
+                        found_similar_failure = True
+                        break
+                if found_similar_failure == False:
+                    failures[job["failureCaptures"]] = [job]
+
+        return failures
+
+    # A flaky job is if it's the only job that has that failureCapture and is not the most recent job
+    def get_flaky_jobs(self, job_statuses: List[Any]) -> List[Any]:
+        unique_failures = self.get_unique_failures(job_statuses)
+        flaky_jobs = []
+        for failure in unique_failures:
+            if len(failure) == 1 and failure["sha"] != self.current_status["sha"]:
+                flaky_jobs.append(failure)
+        return flaky_jobs
+
+    # The most recent failure chain is an array of jobs that have the same-ish failures.
+    # A success in the middle of the chain will terminate the chain.
+    def get_most_recent_failure_chain(self, job_statuses: List[Any]) -> List[Any]:
+        failures = []
+        found_most_recent_failure = False
+
+        for job in job_statuses:
```

Comment:
May be we could make this a bit simpler and returns right away if `job["conclusion"]` is success here because we only alert if `self.current_status["conclusion"] != "success"` anyway (line 119).  So why bother with cases like S-F-F-S-S?  Thoughts?

Reply:
I don't think it's too much extra work to calculate this. We might want to get the most recent failure chain anyways to analysis purposes anyways.